community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/pipes/aggregation/AvgFunction.scala

```diff
@@ -31,23 +31,34 @@ import org.neo4j.values.utils.ValueMath.overflowSafeAdd
  */
 class AvgFunction(val value: Expression)
   extends AggregationFunction
-  with NumericExpressionOnly {
+    with NumericOrDurationAggregationExpression {
 
   def name = "AVG"
 
   private var count: Long = 0L
-  private var sum: NumberValue = Values.ZERO_INT
 
-  override def result(state: QueryState): Value =
-    if (count > 0) sum
-    else Values.NO_VALUE
+  override def result(state: QueryState): Value = aggregatingType match {
+    case None =>
+      Values.NO_VALUE
+    case Some(AggregatingNumbers) =>
+      sumNumber
+    case Some(AggregatingDurations) =>
+      sumDuration.div(Values.longValue(count))
```

Comment:
Doing a sum and then taking the average does not allow for taking the average of many durations where the sum of seconds (or days or months) is greater than what a long can handle.
Should use CMA instead for durations as well.

Reply:
We implemented it that way first. The result is that you get equivalent (but not _equal_) Durations to what you would expect. Example with CMA:
```
avg(P3D, P2DT2H) =
P3D + (P2DT2H - P3D) / 2 =
P3D + P-1DT2H / 2 =
P3D + PT-10H =
P3DT-10H
```

whereas with sum-first-average:
```
avg(P3D, P2DT2H) =
(P3D + P2DT2H) / 2 =
P5DT2H / 2 =
P2DT14H
```

Any idea how to solve it the best way?

Reply:
I guess doing CMA on four different `double` values would work, and then only joining them to a duration in the end.